File: lib/src/data/cache_repository.dart
import 'package:sqlite3/sqlite3.dart';

import 'app_database.dart';

/// A tiny string key/value cache backed by the `cache` table.
///
/// Used to memoize translation and definition lookups so we respect the free
/// APIs' rate limits and so previously looked-up words keep working offline.
class CacheRepository {
  CacheRepository(this._appDb);

  final AppDatabase _appDb;
  Database get _db => _appDb.db;

  String? get(String key) {
    final rows = _db.select(
      'SELECT value FROM cache WHERE cache_key = ? LIMIT 1;',
      [key],
    );
    if (rows.isEmpty) return null;
    return rows.first['value'] as String;
  }

  void put(String key, String value) {
    _db.execute(
      '''
      INSERT INTO cache(cache_key, value, created_at) VALUES(?,?,?)
      ON CONFLICT(cache_key) DO UPDATE SET value = excluded.value, created_at = excluded.created_at;
      ''',
      [key, value, DateTime.now().millisecondsSinceEpoch],
    );
  }

  void clear() => _db.execute('DELETE FROM cache;');
}

File: ios-native/Eng/Data/CacheRepository.swift
import Foundation

/// A tiny key/value cache backed by the `cache` table. Used to memoize
/// translation/definition lookups (provider-namespaced keys) so free quotas are
/// respected and results work offline afterwards.
struct CacheRepository {
    let db: Database

    func get(_ key: String) -> String? {
        (try? db.query("SELECT value FROM cache WHERE cache_key=?;", [key]) { $0.string(0) })?.first
    }

    func put(_ key: String, _ value: String) {
        _ = try? db.run("INSERT OR REPLACE INTO cache(cache_key, value, created_at) VALUES(?,?,?);",
                        [key, value, Date()])
    }
}
